File: analytics-service/app/consumer.py
"""
Analytics & Monitoring Service - Event consumer and statistics.

Consumes ALL journey events from RabbitMQ and:
1. Logs them to the event_logs table for historical analysis
2. Updates hourly aggregate statistics
3. Maintains real-time counters in Redis
"""

import json
import uuid
import os
import logging
from datetime import datetime, timedelta

import redis.asyncio as redis_async
from sqlalchemy import select, func, and_

from .database import EventLog, HourlyStats, async_session
from shared.messaging import MessageBroker, ANALYTICS_QUEUE
from shared.schemas import EventType

logger = logging.getLogger(__name__)

REDIS_URL = os.getenv("REDIS_URL", "redis://redis:6379/5")
redis_client = redis_async.from_url(REDIS_URL, decode_responses=True)


async def handle_event(data: dict, routing_key: str):
    """Process an incoming event for analytics."""
    logger.info(f"Analytics received event: {routing_key}")

    # Store event in database
    async with async_session() as db:
        event = EventLog(
            id=str(uuid.uuid4()),
            event_type=routing_key,
            journey_id=data.get("journey_id"),
            user_id=data.get("user_id"),
            origin=data.get("origin"),
            destination=data.get("destination"),
            metadata_json=json.dumps(data),
        )
        db.add(event)
        await db.commit()

    # Update Redis counters for real-time stats
    try:
        today = datetime.utcnow().strftime("%Y-%m-%d")
        counter_key = f"analytics:daily:{today}"

        pipe = redis_client.pipeline()
        pipe.hincrby(counter_key, "total_events", 1)
        pipe.hincrby(counter_key, routing_key, 1)
        pipe.expire(counter_key, 86400 * 2)  # 2 days TTL
        await pipe.execute()
    except Exception as e:
        logger.warning(f"Failed to update Redis counters: {e}")


async def get_system_stats() -> dict:
    """Get real-time system statistics from Redis + DB."""
    today = datetime.utcnow().strftime("%Y-%m-%d")
    stats = {
        "total_events_today": 0,
        "confirmed_today": 0,
        "rejected_today": 0,
        "cancelled_today": 0,
    }

    try:
        counter_key = f"analytics:daily:{today}"
        data = await redis_client.hgetall(counter_key)
        stats["total_events_today"] = int(data.get("total_events", 0))
        stats["confirmed_today"] = int(data.get(EventType.JOURNEY_CONFIRMED.value, 0))
        stats["rejected_today"] = int(data.get(EventType.JOURNEY_REJECTED.value, 0))
        stats["cancelled_today"] = int(data.get(EventType.JOURNEY_CANCELLED.value, 0))
    except Exception as e:
        logger.warning(f"Failed to get Redis stats: {e}")

    # Get historical stats from DB
    async with async_session() as db:
        result = await db.execute(
            select(func.count()).select_from(EventLog)
        )
        stats["total_events_all_time"] = result.scalar() or 0

        # Events in last hour
        one_hour_ago = datetime.utcnow() - timedelta(hours=1)
        result = await db.execute(
            select(func.count()).select_from(EventLog).where(
                EventLog.created_at >= one_hour_ago
            )
        )
        stats["events_last_hour"] = result.scalar() or 0

    return stats


async def get_event_history(
    event_type: str = None, limit: int = 50, offset: int = 0
) -> list[dict]:
    """Get event history with optional filtering."""
    async with async_session() as db:
        query = select(EventLog).order_by(EventLog.created_at.desc())
        if event_type:
            query = query.where(EventLog.event_type == event_type)
        query = query.offset(offset).limit(limit)

        result = await db.execute(query)
        events = result.scalars().all()

        return [
            {
                "id": e.id,
                "event_type": e.event_type,
                "journey_id": e.journey_id,
                "user_id": e.user_id,
                "origin": e.origin,
                "destination": e.destination,
                "created_at": e.created_at.isoformat(),
            }
            for e in events
        ]


async def start_consumer(broker: MessageBroker):
    """Start consuming ALL journey events for analytics."""
    await broker.subscribe(
        queue_name=ANALYTICS_QUEUE,
        routing_keys=[
            "journey.*",   # All journey events
            "user.*",      # All user events
        ],
        callback=handle_event,
    )
    logger.info("Analytics consumer started")

File: analytics-service/consumer.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"time"

	amqp "github.com/rabbitmq/amqp091-go"
	"github.com/google/uuid"
	"github.com/redis/go-redis/v9"
)

const (
	eventsExchange = "journey_events"
	dlxExchange    = "journey_events_dlx"
	analyticsQueue = "analytics_events"
)

var redisClient *redis.Client

func initRedis(redisURL string) error {
	opts, err := redis.ParseURL(redisURL)
	if err != nil {
		return fmt.Errorf("invalid Redis URL: %w", err)
	}
	redisClient = redis.NewClient(opts)
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	return redisClient.Ping(ctx).Err()
}

func startConsumer(rabbitURL string) error {
	var conn *amqp.Connection
	var err error

	for attempt := 1; attempt <= 10; attempt++ {
		conn, err = amqp.Dial(rabbitURL)
		if err == nil {
			break
		}
		log.Printf("RabbitMQ connection attempt %d/10 failed: %v", attempt, err)
		if attempt < 10 {
			time.Sleep(3 * time.Second)
		}
	}
	if err != nil {
		return fmt.Errorf("failed to connect to RabbitMQ after 10 attempts: %w", err)
	}

	if err := setupAnalyticsChannel(conn); err != nil {
		conn.Close()
		return err
	}

	// Reconnect on connection drop
	go func() {
		errCh := conn.NotifyClose(make(chan *amqp.Error, 1))
		<-errCh
		log.Println("RabbitMQ connection lost, reconnecting...")
		for {
			time.Sleep(3 * time.Second)
			if err := startConsumer(rabbitURL); err != nil {
				log.Printf("Reconnect failed: %v", err)
			} else {
				log.Println("Reconnected to RabbitMQ")
				return
			}
		}
	}()

	return nil
}

func setupAnalyticsChannel(conn *amqp.Connection) error {
	ch, err := conn.Channel()
	if err != nil {
		return err
	}
	if err := ch.Qos(10, 0, false); err != nil {
		return err
	}

	// Dead-letter exchange
	if err := ch.ExchangeDeclare(dlxExchange, "fanout", true, false, false, false, nil); err != nil {
		return err
	}
	dlq, err := ch.QueueDeclare("dead_letter_queue", true, false, false, false, nil)
	if err != nil {
		return err
	}
	if err := ch.QueueBind(dlq.Name, "", dlxExchange, false, nil); err != nil {
		return err
	}

	// Main topic exchange
	if err := ch.ExchangeDeclare(eventsExchange, "topic", true, false, false, false, nil); err != nil {
		return err
	}

	q, err := ch.QueueDeclare(analyticsQueue, true, false, false, false, amqp.Table{
		"x-dead-letter-exchange": dlxExchange,
		"x-message-ttl":          int32(86400000), // 24h TTL
	})
	if err != nil {
		return err
	}

	// Subscribe to all journey.* and user.* events
	for _, key := range []string{"journey.*", "user.*"} {
		if err := ch.QueueBind(q.Name, key, eventsExchange, false, nil); err != nil {
			return err
		}
		log.Printf("Queue '%s' bound to routing key '%s'", q.Name, key)
	}

	msgs, err := ch.Consume(q.Name, "", false, false, false, false, nil)
	if err != nil {
		return err
	}

	log.Println("Analytics consumer started")
	go func() {
		for msg := range msgs {
			if err := handleEvent(msg.Body, msg.RoutingKey); err != nil {
				log.Printf("Error processing message: %v", err)
				msg.Nack(false, false)
			} else {
				msg.Ack(false)
			}
		}
		log.Println("RabbitMQ channel closed")
	}()

	return nil
}

func handleEvent(body []byte, routingKey string) error {
	log.Printf("Analytics received event: %s", routingKey)

	var data map[string]any
	if err := json.Unmarshal(body, &data); err != nil {
		return fmt.Errorf("unmarshal event: %w", err)
	}

	getString := func(key string) string {
		if v, ok := data[key].(string); ok {
			return v
		}
		return ""
	}

	metaJSON, _ := json.Marshal(data)

	if err := insertEvent(
		context.Background(),
		uuid.New().String(),
		routingKey,
		getString("journey_id"),
		getString("user_id"),
		getString("origin"),
		getString("destination"),
		string(metaJSON),
	); err != nil {
		log.Printf("Failed to insert event to DB: %v", err)
	}

	// Update Redis daily counters (best-effort)
	if redisClient != nil {
		today := time.Now().UTC().Format("2006-01-02")
		key := "analytics:daily:" + today
		ctx := context.Background()
		pipe := redisClient.Pipeline()
		pipe.HIncrBy(ctx, key, "total_events", 1)
		pipe.HIncrBy(ctx, key, routingKey, 1)
		pipe.Expire(ctx, key, 48*time.Hour)
		if _, err := pipe.Exec(ctx); err != nil {
			log.Printf("Failed to update Redis counters: %v", err)
		}
	}

	return nil
}

// GetSystemStats returns a mix of real-time Redis counters and DB aggregates.
func GetSystemStats(ctx context.Context) map[string]any {
	stats := map[string]any{
		"total_events_today": 0,
		"confirmed_today":    0,
		"rejected_today":     0,
		"cancelled_today":    0,
		"total_events_all_time": 0,
		"events_last_hour":   0,
	}

	if redisClient != nil {
		today := time.Now().UTC().Format("2006-01-02")
		key := "analytics:daily:" + today
		data, err := redisClient.HGetAll(ctx, key).Result()
		if err == nil {
			stats["total_events_today"] = atoi(data["total_events"])
			stats["confirmed_today"]    = atoi(data["journey.confirmed"])
			stats["rejected_today"]     = atoi(data["journey.rejected"])
			stats["cancelled_today"]    = atoi(data["journey.cancelled"])
		}
	}

	if total, err := getTotalEvents(ctx); err == nil {
		stats["total_events_all_time"] = total
	}
	if lastHour, err := getEventsLastHour(ctx); err == nil {
		stats["events_last_hour"] = lastHour
	}

	return stats
}

func atoi(s string) int64 {
	var n int64
	fmt.Sscanf(s, "%d", &n)
	return n
}
